Rewrite Svelte adapter with Svelte v5 syntax

Related to #323

Try it after v5 becomes popular enough.

@inokawa seems like `virtua` can't be installed for Svelte 5 project due to dependency conflict, because it expects only version 4 in package.json

```
❯ npm install virtua
npm error code ERESOLVE
npm error ERESOLVE could not resolve
npm error
npm error While resolving: virtua@0.35.0
npm error Found: svelte@5.0.0-next.263
npm error node_modules/svelte
npm error   peer svelte@"^5.0.0-next.1" from @skeletonlabs/floating-ui-svelte@0.3.1
npm error   node_modules/@skeletonlabs/floating-ui-svelte
npm error     @skeletonlabs/floating-ui-svelte@"^0.3.1" from the root project
npm error   peer svelte@"^4.0.0 || ^5.0.0-next.0" from @sveltejs/kit@2.6.3
npm error   node_modules/@sveltejs/kit
npm error     peer @sveltejs/kit@"^2.0.0" from @sveltejs/adapter-auto@3.2.5
npm error     node_modules/@sveltejs/adapter-auto
npm error       dev @sveltejs/adapter-auto@"^3.2.3" from the root project
npm error     peer @sveltejs/kit@"^2.0.0" from @sveltejs/adapter-static@3.0.5
npm error     node_modules/@sveltejs/adapter-static
npm error       dev @sveltejs/adapter-static@"^3.0.3" from the root project
npm error     1 more (the root project)
npm error   7 more (@sveltejs/vite-plugin-svelte, ...)
npm error
npm error Could not resolve dependency:
npm error peerOptional svelte@">=4.0" from virtua@0.35.0
npm error node_modules/virtua
npm error   virtua@"*" from the root project
npm error
npm error Conflicting peer dependency: svelte@4.2.19
npm error node_modules/svelte
npm error   peerOptional svelte@">=4.0" from virtua@0.35.0
npm error   node_modules/virtua
npm error     virtua@"*" from the root project
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
```

--- a/README.md
+++ b/README.md
@@ -252,7 +252,7 @@ export const App = () => {
 
 ### Svelte
 
-`svelte >= 4.0` is required.
+`svelte >= 5.0` is required.
 
 ```svelte
 <script lang="ts">
@@ -263,16 +263,18 @@ export const App = () => {
   const data = Array.from({ length: 1000 }).map((_, i) => sizes[i % 4] );
 </script>
 
-<VList {data} let:item let:index style={`height: 100vh;`} getKey={(_,i) => i}>
-  <div
-    style={`
-      height: ${item}px;
-      background: white;
-      border-bottom: solid 1px #ccc;
-    `}
-  >
-    {index}
-  </div>
+<VList {data} style={`height: 100vh;`} getKey={(_, i) => i}>
+  {#snippet children({ item, index })}
+    <div
+      style={`
+        height: ${item}px;
+        background: white;
+        border-bottom: solid 1px #ccc;
+      `}
+    >
+      {index}
+    </div>
+  {/snippet}
 </VList>
 ```
 

--- a/package.json
+++ b/package.json
@@ -86,7 +86,7 @@
     "@storybook/test-runner": "^0.19.1",
     "@storybook/vue3": "^8.3.5",
     "@storybook/vue3-vite": "^8.3.5",
-    "@sveltejs/vite-plugin-svelte": "^3.1.2",
+    "@sveltejs/vite-plugin-svelte": "^4.0.0",
     "@tanstack/react-virtual": "^3.10.8",
     "@testing-library/dom": "^10.4.0",
     "@testing-library/react": "^16.0.1",
@@ -133,8 +133,8 @@
     "storybook": "^8.3.5",
     "storybook-solidjs": "^1.0.0-beta.2",
     "storybook-solidjs-vite": "^1.0.0-beta.2",
-    "svelte": "^4.2.19",
-    "svelte-check": "^3.8.6",
+    "svelte": "^5.0.0",
+    "svelte-check": "^4.0.0",
     "svelte2tsx": "^0.7.22",
     "typedoc": "^0.26.10",
     "typedoc-plugin-markdown": "^4.2.9",
@@ -152,7 +152,7 @@
     "react": ">=16.14.0",
     "react-dom": ">=16.14.0",
     "solid-js": ">=1.0",
-    "svelte": ">=4.0",
+    "svelte": ">=5.0",
     "vue": ">=3.2"
   },
   "peerDependenciesMeta": {

--- a/src/svelte/ListItem.svelte
+++ b/src/svelte/ListItem.svelte
@@ -1,20 +1,27 @@
-<script lang="ts">
-  import { afterUpdate, onDestroy } from "svelte";
+<script lang="ts" generics="T">
+  import { type Snippet, onDestroy } from "svelte";
   import { isRTLDocument, type ItemResizeObserver } from "./core";
   import { styleToString } from "./utils";
 
-  export let index: number;
-  export let offset: number;
-  export let hide: boolean;
-  export let horizontal: boolean;
-  export let resizer: ItemResizeObserver;
+  interface Props {
+    children: Snippet<[{ item: T; index: number }]>;
+    item: T;
+    index: number;
+    offset: number;
+    hide: boolean;
+    horizontal: boolean;
+    resizer: ItemResizeObserver;
+  }
+
+  let { children, item, index, offset, hide, horizontal, resizer }: Props =
+    $props();
 
   let elementRef: HTMLDivElement;
 
   let cleanupResizer: (() => void) | undefined;
   // The index may be changed if elements are inserted to or removed from the start of props.children
   let prevIndex: number | undefined;
-  afterUpdate(() => {
+  $effect(() => {
     if (prevIndex === index) return;
     if (cleanupResizer) cleanupResizer();
     cleanupResizer = resizer(elementRef, (prevIndex = index));
@@ -23,8 +30,7 @@
     if (cleanupResizer) cleanupResizer();
   });
 
-  let style: string;
-  $: {
+  let style: string = $derived.by(() => {
     const _style: Record<string, string> = {
       position: "absolute",
       [horizontal ? "height" : "width"]: "100%",
@@ -37,10 +43,10 @@
       _style["display"] = "flex";
     }
 
-    style = styleToString(_style);
-  }
+    return styleToString(_style);
+  });
 </script>
 
 <div bind:this={elementRef} {style}>
-  <slot />
+  {@render children({ item, index })}
 </div>

--- a/src/svelte/VList.svelte
+++ b/src/svelte/VList.svelte
@@ -1,6 +1,5 @@
 <script lang="ts" generics="T">
-  import { onMount, onDestroy, createEventDispatcher } from "svelte";
-  import type { HTMLAttributes } from "svelte/elements";
+  import { type Snippet, onMount, onDestroy } from "svelte";
   import {
     SCROLL_IDLE,
     type StateVersion,
@@ -24,23 +23,19 @@
     SCROLL_TO,
     SCROLL_TO_INDEX,
   } from "./core";
-  import {
-    onUpdate,
-    type UnwrapCustomEvents,
-    defaultGetKey,
-    styleToString,
-  } from "./utils";
+  import { defaultGetKey, styleToString } from "./utils";
   import ListItem from "./ListItem.svelte";
+  import type { ViewportComponentAttributes } from "./types";
 
-  // https://github.com/sveltejs/rfcs/pull/38
-  interface $$Slots {
-    default: { item: T; index: number };
-  }
-  interface $$Props extends HTMLAttributes<HTMLDivElement> {
+  interface Props extends ViewportComponentAttributes {
     /**
      * The data items rendered by this component.
      */
     data: T[];
+    /**
+     * The elements renderer snippet.
+     */
+    children: Snippet<[{ item: T; index: number }]>;
     /**
      * Function that returns the key of an item in the list. It's recommended to specify whenever possible for performance.
      * @default defaultGetKey (returns index of item)
@@ -66,37 +61,47 @@
      * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
      */
     horizontal?: boolean;
-  }
-
-  // https://github.com/kitschpatrol/svelte-tweakpane-ui/blob/248fd7c24926af1e485a0676b6a8c053177e92db/src/lib/control/Button.svelte#L50-L51
-  interface $$Events {
     /**
      * Callback invoked whenever scroll offset changes.
      * @param offset Current scrollTop or scrollLeft.
      */
-    scroll: CustomEvent<number>;
+    onscroll?: (offset: number) => void;
     /**
      * Callback invoked when scrolling stops.
      */
-    scrollEnd: CustomEvent<void>;
+    onscrollend?: () => void;
     /**
      * Callback invoked when visible items range changes.
      */
-    rangeChange: CustomEvent<[start: number, end: number]>;
+    onrangechange?: (
+      /**
+       * The start index of viewable items.
+       */
+      startIndex: number,
+      /**
+       * The end index of viewable items.
+       */
+      endIndex: number
+    ) => void;
   }
 
-  export let data: $$Props["data"];
-  export let getKey: NonNullable<$$Props["getKey"]> = defaultGetKey;
-  export let overscan: NonNullable<$$Props["overscan"]> = 4;
-  export let itemSize: $$Props["itemSize"] = undefined;
-  export let shift: NonNullable<$$Props["shift"]> = false;
-  export let horizontal: NonNullable<$$Props["horizontal"]> = false;
+  let {
+    data,
+    getKey = defaultGetKey,
+    overscan = 4,
+    itemSize,
+    shift = false,
+    horizontal = false,
+    children,
+    onscroll,
+    onscrollend,
+    onrangechange,
+    ...rest
+  }: Props = $props();
 
-  let containerRef: HTMLDivElement;
+  let containerRef: HTMLDivElement | undefined = $state();
 
-  let rerender: StateVersion = [];
-
-  const dispatchEvent = createEventDispatcher<UnwrapCustomEvents<$$Events>>();
+  let rerender: StateVersion = $state([]);
 
   const virtualizer = createVirtualizer(
     data.length,
@@ -106,27 +111,31 @@
       rerender = v;
     },
     (offset) => {
-      dispatchEvent("scroll", offset);
+      onscroll && onscroll(offset);
     },
     () => {
-      dispatchEvent("scrollEnd");
+      onscrollend && onscrollend();
     }
   );
 
-  $: range = rerender && virtualizer[GET_RANGE]();
-  $: scrollDirection = rerender && virtualizer[GET_SCROLL_DIRECTION]();
-  $: totalSize = rerender && virtualizer[GET_TOTAL_SIZE]();
-  $: jumpCount = rerender && virtualizer[GET_JUMP_COUNT]();
-  $: extendedRange = getOverscanedRange(
-    range[0],
-    range[1],
-    overscan,
-    scrollDirection,
-    data.length
+  let range = $derived(rerender && virtualizer[GET_RANGE]());
+  let scrollDirection = $derived(
+    rerender && virtualizer[GET_SCROLL_DIRECTION]()
+  );
+  let totalSize = $derived(rerender && virtualizer[GET_TOTAL_SIZE]());
+  let jumpCount = $derived(rerender && virtualizer[GET_JUMP_COUNT]());
+  let extendedRange = $derived(
+    getOverscanedRange(
+      range[0],
+      range[1],
+      overscan,
+      scrollDirection,
+      data.length
+    )
   );
 
   onMount(() => {
-    const root = containerRef.parentElement!;
+    const root = containerRef!.parentElement!;
     virtualizer[ON_MOUNT](root);
   });
   onDestroy(() => {
@@ -134,24 +143,24 @@
   });
 
   let prevLength = data.length;
-  onUpdate(() => {
+  $effect.pre(() => {
     if (prevLength === data.length) return;
     virtualizer[CHANGE_ITEM_LENGTH]((prevLength = data.length), shift);
-  }, true);
+  });
 
   let prevJumpCount: number | undefined;
-  onUpdate(() => {
+  $effect(() => {
     if (prevJumpCount === jumpCount) return;
     prevJumpCount = jumpCount;
     virtualizer[FIX_SCROLL_JUMP]();
   });
 
   let prevRange: typeof range | undefined;
-  onUpdate(() => {
+  $effect(() => {
     if (prevRange && prevRange[0] === range[0] && prevRange[1] === range[1])
       return;
     prevRange = range;
-    dispatchEvent("rangeChange", [range[0], range[1]]);
+    onrangechange && onrangechange(range[0], range[1]);
   });
 
   /**
@@ -183,15 +192,14 @@
    */
   export const scrollBy = virtualizer[SCROLL_BY];
 
-  let items: T[];
-  $: {
+  let items = $derived.by(() => {
     const [startIndex, endIndex] = extendedRange;
     const newItems: T[] = [];
     for (let i = startIndex, j = endIndex; i <= j; i++) {
       newItems.push(data[i]!);
     }
-    items = newItems;
-  }
+    return newItems;
+  });
 
   const viewportStyle = styleToString({
     display: horizontal ? "inline-block" : "block",
@@ -209,18 +217,20 @@
     visibility: "hidden", // TODO replace with other optimization methods
   });
 
-  $: dynamicContainerStyle = styleToString({
-    width: horizontal ? totalSize + "px" : "100%",
-    height: horizontal ? "100%" : totalSize + "px",
-    "pointer-events": scrollDirection !== SCROLL_IDLE ? "none" : undefined,
-  });
+  let dynamicContainerStyle = $derived(
+    styleToString({
+      width: horizontal ? totalSize + "px" : "100%",
+      height: horizontal ? "100%" : totalSize + "px",
+      "pointer-events": scrollDirection !== SCROLL_IDLE ? "none" : undefined,
+    })
+  );
 </script>
 
 <!-- 
   @component
   Virtualized list component.
 -->
-<div {...$$restProps} style={`${viewportStyle} ${$$restProps["style"] || ""}`}>
+<div {...rest} style={`${viewportStyle} ${rest["style"] || ""}`}>
   <div
     bind:this={containerRef}
     style={`${containerStyle} ${dynamicContainerStyle}`}
@@ -228,14 +238,14 @@
     {#each items as item, i (getKey(item, i + extendedRange[0]))}
       {@const index = i + extendedRange[0]}
       <ListItem
+        {children}
+        {item}
         {index}
         offset={rerender && virtualizer[GET_ITEM_OFFSET](index)}
         hide={rerender && virtualizer[IS_ITEM_HIDDEN](index)}
         {horizontal}
         resizer={virtualizer[OBSERVE_ITEM_RESIZE]}
-      >
-        <slot {item} {index} />
-      </ListItem>
+      />
     {/each}
   </div>
 </div>

--- a/src/svelte/types.ts
+++ b/src/svelte/types.ts
@@ -1,0 +1,7 @@
+import type { HTMLAttributes, AriaAttributes } from "svelte/elements";
+
+export type ViewportComponentAttributes = Pick<
+  HTMLAttributes<HTMLElement>,
+  "class" | "style" | "id" | "role" | "tabindex"
+> &
+  AriaAttributes;

--- a/src/svelte/utils.ts
+++ b/src/svelte/utils.ts
@@ -1,11 +1,3 @@
-import { afterUpdate, beforeUpdate } from "svelte";
-
-export const onUpdate = (cb: () => void, before?: boolean) => {
-  (before ? beforeUpdate : afterUpdate)(() => {
-    cb();
-  });
-};
-
 export const styleToString = (
   obj: Record<string, string | undefined>
 ): string => {
@@ -19,10 +11,3 @@ export const styleToString = (
 };
 
 export const defaultGetKey = (_data: unknown, i: number) => "_" + i;
-
-/**
- * https://github.com/kitschpatrol/svelte-tweakpane-ui/blob/248fd7c24926af1e485a0676b6a8c053177e92db/src/lib/utils.ts#L57-L59
- */
-export type UnwrapCustomEvents<T> = {
-  [P in keyof T]: T[P] extends CustomEvent<infer detail> ? detail : never;
-};

--- a/stories/svelte/Controls.svelte
+++ b/stories/svelte/Controls.svelte
@@ -4,11 +4,11 @@
   const sizes = [20, 40, 180, 77];
   const createItem = (i: number) => ({ id: i, size: sizes[i % 4] + "px" });
 
-  let ref: VList<{ id: number; size: string }>;
-  let data = Array.from({ length: 1000 }).map((_, i) => createItem(i));
-  let scrollOffset = 0;
-  let scrolling = false;
-  let scrollTarget = 567;
+  let ref: VList<{ id: number; size: string }> = $state();
+  let data = $state(Array.from({ length: 1000 }).map((_, i) => createItem(i)));
+  let scrollOffset = $state(0);
+  let scrolling = $state(false);
+  let scrollTarget = $state(567);
 </script>
 
 <div style="height: 100%; display: flex; flex-direction: column;">
@@ -18,19 +18,19 @@
     <input
       type="number"
       bind:value={scrollTarget}
-      on:input={(e) => {
-        scrollTarget = Number(e.target.value);
+      oninput={(e) => {
+        scrollTarget = Number(e.currentTarget.value);
       }}
     />
     <button
-      on:click={() => {
+      onclick={() => {
         ref.scrollToIndex(scrollTarget);
       }}>scrollToIndex</button
     >
   </div>
   <div>
     <button
-      on:click={() => {
+      onclick={() => {
         data = [
           ...data,
           ...Array.from({ length: 100 }).map((_, i) =>
@@ -43,24 +43,25 @@
   <VList
     bind:this={ref}
     {data}
-    let:item
     getKey={(d) => d.id}
-    on:scroll={(event) => {
-      scrollOffset = event.detail;
+    onscroll={(offset) => {
+      scrollOffset = offset;
       scrolling = true;
     }}
-    on:scrollEnd={() => {
+    onscrollend={() => {
       scrolling = false;
     }}
   >
-    <div
-      style={`
-      height: ${item.size};
-      background: white;
-      border-bottom: solid 1px #ccc;
-    `}
-    >
-      {item.id}
-    </div>
+    {#snippet children({ item })}
+      <div
+        style={`
+          height: ${item.size};
+          background: white;
+          border-bottom: solid 1px #ccc;
+        `}
+      >
+        {item.id}
+      </div>
+    {/snippet}
   </VList>
 </div>

--- a/stories/svelte/Default.svelte
+++ b/stories/svelte/Default.svelte
@@ -6,14 +6,16 @@
   const data = Array.from({ length: 1000 }).map((_, i) => sizes[i % 4]!);
 </script>
 
-<VList {data} let:item let:index style={`height: 100vh;`} getKey={(_, i) => i}>
-  <div
-    style={`
-      height: ${item}px;
-      background: white;
-      border-bottom: solid 1px #ccc;
-    `}
-  >
-    {index}
-  </div>
+<VList {data} style={`height: 100vh;`} getKey={(_, i) => i}>
+  {#snippet children({ item, index })}
+    <div
+      style={`
+        height: ${item}px;
+        background: white;
+        border-bottom: solid 1px #ccc;
+      `}
+    >
+      {index}
+    </div>
+  {/snippet}
 </VList>

--- a/stories/svelte/Horizontal.svelte
+++ b/stories/svelte/Horizontal.svelte
@@ -10,19 +10,20 @@
 <div style="padding: 10px;">
   <VList
     {data}
-    let:item
     style={`width: 100%; height: 200px;`}
     getKey={(d) => d.id}
     horizontal
   >
-    <div
-      style={`
-      width: ${item.size};
-      background: white;
-      border-right: solid 1px #ccc;
-    `}
-    >
-      {item.id}
-    </div>
+    {#snippet children({ item })}
+      <div
+        style={`
+          width: ${item.size};
+          background: white;
+          border-right: solid 1px #ccc;
+        `}
+      >
+        {item.id}
+      </div>
+    {/snippet}
   </VList>
 </div>

